Assets/Scripts/KatamariMovement.cs

```diff
@@ -314,205 +297,168 @@ void OptimizeOldestStuckObjects()
-                myStateMachine = StateMachine.slowDown;
+                if (Time.timeScale == slowdownFactor)
+                {
+                    movementState = StateMachine.slowMotion;
 
-            }
-            if (strokeDone) {
-                myStateMachine = StateMachine.toGolfMode;
-            }
+                    PointArrowAwayFromKatamari();
+                    arrow.SetActive(true);
+                }
 
-            if (Time.timeScale == 1) {
-                //Transition Back to Normal Speed\\
-                myStateMachine = StateMachine.normalSpeed;
-                flyStrength = 0;
-            }
+                if (!isHitInputActive)
+                {
+                    movementState = StateMachine.speedUp;
+                    arrow.SetActive(false); //maybe
+                }
+
+                break;
+
+            case StateMachine.slowMotion:
+                if (!isHitInputActive)
+                {
+                    movementState = StateMachine.speedUp;
+                    arrow.SetActive(false); // maybe
+                }
+
+                break;
+
+            case StateMachine.speedUp:
+                IncreaseTimeScale();
+
+                // Scale back the velocity from slow motion to prevent unexpected momentum
+                rb.velocity *= .98f;
+
+                if (isHitInputActive)
+                {
+                    movementState = StateMachine.slowDown;
+                }
+
+                if (Time.timeScale == 1)
+                {
+                    movementState = StateMachine.normalSpeed;
+                }
+                break;
+
+            default:
+                print("Unrecognized state reached: " + movementState);
+                break;
         }
     }
 
     public virtual void OnCollisionStay(Collision collision)
     {
+        // DrawTrajectory simulation may run this before Start() is called, so the RigidBody
+        // component reference may not have happened yet
+        if (rb is null) { rb = GetComponent<Rigidbody>(); }
 
-        if (myStateMachine != StateMachine.golfMode && myStateMachine != StateMachine.slowMotion && collision.gameObject.tag =="green")
+        if (movementState != StateMachine.golfMode && movementState != StateMachine.slowMotion && collision.gameObject.tag == "green")
         {
             timeOnGround += Time.deltaTime;
-            print(timeOnGround);
-
             rb.drag += 0.005f;
             rb.angularDrag += .2f;
-         
-            
-            
-
-        } else if (myStateMachine == StateMachine.golfMode || myStateMachine == StateMachine.slowMotion)
+        }
+        else if (movementState == StateMachine.golfMode || movementState == StateMachine.slowMotion)
         {
             rb.drag = 0f;
             rb.angularDrag = 0f;
-          
-            
-        } 
-        
-            if (timeOnGround >= timeToStop && rb.velocity.magnitude <= 0.5f)
-            {
-                strokeDone = true;
```

Comment:
I thought this logic made more sense to move into the state machine stuff rather than to signal a state transition from the collider callback, so I moved it into the normal speed state update.